### PR TITLE
Feature to development pull request

### PR DESCRIPTION
**Pull request by Daniduran-dev**
In this project I have replicated some of the most common enumerable Ruby methods.
You will find replicas for #each, #each_with_index, #count, #all, #any, #none, #map, #select and #inject.

- I have been trying to fix the rubocop but it's impossible. I keep fixing it and it tells me theres new errors, then i fix those errors and tells me i should  re-fix the previous errors. It's a loop impossible to break.

### CODE REVIEW

**Comment by Daniduran-dev** [comment]
Ok.. rubocop in Github doesn't show errors! Finally!

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    size.times do |x|
+      yield(self[x])
+    end
+    self
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_each
1. [x] When a block is given when self is a range, returns the range itself after calling the given block once for each element in self
### Test 👉 `range.my_each(&block)`
Resulted in *NoMethodError* 
#
2. [x] When a block is given when self is a range behaves like #each calls the given block once for each element in self
### Test 👉 `enum.my_each(&block)`

Resulted in *NoMethodError* 
#
3. [x] When a block is given when self is a hash behaves like #each calls the given block once for each element in self
### Test 👉 `my_each_output`

|Expected|Got|
|:---:|:---:|
|`"[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"`|`" "`|

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+  def my_each_with_index(value = 0)
+    return to_enum(:my_each_with_index) unless block_given?
+
+    index = value
+    size.times do |x|
+      yield(self[x], index)
+      index += 1
+    end
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_each_with_index
1. [ ] When a block is given when self is an array returns the array itself after calling the given block once for each element in self
### Test 👉 `array.my_each_with_index(&block)`

|Expected|Got|
|:---:|:---:|
|`[6, 0, 4, 8, 7, 7, 5, 7, 1, 1, 4, 7, 8, 7, 6, 3, 2, 8, 0, 2, 7, 4, 0, 4, 0, 7, 3, 6, 7, 2, 3, 4, 4, 3..., 6, 3, 5, 6, 6, 5, 1, 0, 7, 8, 5, 1, 6, 2, 4, 5, 5, 8, 8, 3, 2, 2, 3, 6, 3, 8, 8, 7, 1, 1, 4, 8, 5]`|`100`|
# 
2. [ ] When a block is given when self is a range returns the range itself after calling the given block once for each element in self
### Test 👉 `range.my_each_with_index(&block)`
Resulted in *NoMethodError*
#
3. When a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
### Test 👉 `enum.my_each_with_index(&block)`
Resulted in *NoMethodError*
#
4. When a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
### Test 👉 `my_each_with_index_output`

|Expected|Got|
|:---:|:---:|
|`"Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"`|`"Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"`|
#

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      my_each do |x|
+        arr.push(x) if arg(x)
+      end
+    else
+      my_each do |x|
+        arr.push(x) if yield(x)
+      end
+    end
+    arr
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_select
1. [ ] Returns an array containing all elements of enum for which the given block returns a true value:: range
### Test 👉 `range.my_select(&block)`
Resulted in *NoMethodError*

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      my_each do |x|
+        return false unless x.class == arg
+      end
+    else
+      my_each do |x|
+        return false unless arg == x
+      end
+    end
+    true
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_all?
1. [ ] Returns false if the block returns false or nil:: range
### Test 👉 `range.my_all?(&false_block)`
Resulted in *NoMethodError* 
#
2. [ ] When no block or argument is given returns true when none of the collection members is false or nil
### Test 👉 `true_array.my_all?`

|Expected|Got|
|:---:|:---:|
|`true`|`false`|
#
3. [ ] When a class is passed as an argument returns true if all of the collection is a member of such class:: Integer
### Test 👉 `array.my_all?(Integer)`

|Expected|Got|
|:---:|:---:|
|`true`|`false`| 
#
4. [ ] When a class is passed as an argument returns true if all of the collection is a member of such class:: Numeric
### Test 👉 `numbers.my_all?(Numeric)`

|Expected|Got|
|:---:|:---:|
|`true`|`false`|

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      my_each do |x|
+        return true if arg == x
+      end
+    else
+      my_each do |x|
+        return true if x.class == arg
+      end
+      false
+    end
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_any?
1. [ ] Returns false if the block ever returns a value false or nil
### Test 👉 `array.my_any?(&false_block)`

|Expected|Got|
|:---:|:---:|
|`false`|` [8, 3, 7, 7, 3, 4, 7, 6, 2, 5, 0, 3, 6, 8, 2, 1, 8, 6, 3, 8, 3, 8, 3, 4, 2, 4, 4, 4, 5, 6, 7, 5, 2, 8..., 4, 4, 1, 6, 2, 5, 0, 4, 3, 4, 1, 1, 5, 3, 1, 0, 7, 6, 1, 3, 0, 7, 2, 0, 2, 2, 0, 2, 2, 0, 6, 7, 6]`| 
#
2. [ ] Returns false if the block ever returns a value false or nil:: range
### Test 👉 `range.my_any?(&false_block)`

Resulted in *NoMethodError*
#
3. [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class:: Numeric

|Expected|Got|
|:---:|:---:|
|`true`|`false`| 
#
4. [ ] When a Regex is passed as an argument returns false if none of the collection matches the Regex
### Test 👉 `words.my_any?(/z/)`

|Expected|Got|
|:---:|:---:|
|`false`|`#<Array:2080> => ["dog", "door", "rod", "blade"]`| 
#
5. When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
### Test 👉 `words.my_any?('cat')`

|Expected|Got|
|:---:|:---:|
|`false`|`#<Array:2080> => ["dog", "door", "rod", "blade"]`|

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                     0
+                   end
+      end
+    else
+      my_each do |_x|
+        counter += 1
+      end
+    end
+    counter
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_none?
1. [ ]  Returns true if the block never returns true for all elements
### Test 👉 `array.my_none?(&true_block)`

|Expected|Got|
|:---:|:---:|
|`true`|`false`| 
#
2. [ ] Returns false if the block ever returns true for all elements:: range
### Test 👉 `array.my_none?(&true_block)`

Resulted in *NoMethodError*
#
3. [ ] When a Regex is passed as an argument returns true if none of the collection matches the Regex
### Test 👉 `words.my_none?(/z/)`

|Expected|Got|
|:---:|:---:|
|`true`|`false`| 
#
4. [ ] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
### Test 👉 `words.my_none?(5)`
|Expected|Got|
|:---:|:---:|
|`true`|`false`|

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      end
+    elsif arg.class == Symbol
+      acc = arr[0]
+      arr[1..arr.length - 1].my_each do |x|
+        acc = acc.send(arg, x)
+      end
+    end
+    acc
+  end
+end
```
Comment (christianotieno) [inline enumerables.rb]: # my_inject
1. [ ]  Raises a "LocalJumpError" when no block or argument is given
### Test 👉 `array.my_inject`
 expected `LocalJumpError` but nothing was raised

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+# rubocop:enable Metrics/ModuleLength, Metrics/PerceivedComplexity, Metrics/MethodLength
+def multiply_els(arg)
+  arg.my_inject(:*)
+end
```
Comment (christianotieno) [inline enumerables.rb:218]: Please Implement `my_count` as well.

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @Daniduran-dev  :wave:
-
You have done a splendid job! :tada: Keep it up.
We, however, do have some few changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Tests Done: 79
## Failures: 23
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Comment by Daniduran-dev** [comment]
Hello,
I'm doing this exercise alone on my own, not with a partner, My teammates gave me the list of the test I should pass and I passed them all. Nobody has ever said anything about hashes before. Can you please apply the old requirements? If I had a partner I wouldn't mind doing it but alone it is going to require more time than I have. Thank you.

**Comment by lekegitrepo** [comment]
> Hello,
> I'm doing this exercise alone on my own, not with a partner, My teammates gave me the list of the test I should pass and I passed them all. Nobody has ever said anything about hashes before. Can you please apply the old requirements? If I had a partner I wouldn't mind doing it but alone it is going to require more time than I have. Thank you.

Hi, as much as I would have like to help, however, the requirements for this project must be met (.i.e all test cases for each method must be passed) before the project can be approved.

**Comment by lekegitrepo** [comment]
#### Hi @Daniduran-dev :clap: :clap:
### You have put a lot of effort on this project but you have to implement the required changes mentioned in the previous review before this project can be approved. Please :pray: complete the required changes mentioned by the previous reviewer.

#### After completing the required changes mentioned in the previous review; you can submit for another review.

#### _Reviewed by Temitayo Adeleke_.

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### Status: Request changes

**Pull request closed without merging**